typing/ctype.ml

```diff
@@ -1756,8 +1760,20 @@ let enforce_constraints env ty =
 
 (* Recursively expand the head of a type.
    Also expand #-types. *)
-let full_expand env ty =
-  let ty = repr (expand_head env ty) in
+let full_expand ~may_forget_scope env ty =
+  let ty =
+    if may_forget_scope then
```

Comment:
I wonder if it would be slightly simpler to pass `may_forget_scope` down to `expand_abbrev_gen` and simply ignore the `type_expansion_scope` when it is true. Perhaps rename the parameter to `ignore_scopes` to make it clearer if you do.

Reply:
I looked into it, but the machinery around `expand_head` is complicated, so that it would be hard to do.
Moreover, I would be concerned by allowing "wrong" handling of potentially shared types (well, we are reporting an error, so this should be fine (either fatal failure, or backtrack in the toplevel), but morally this is still strange).
Since cost is not a problem here, calling `correct_levels` is morally cleaner. We may lose some sharing, but then the goal here is just to show that there is an equation, but we are not allowed to use it.